### PR TITLE
Sprint contains today

### PR DESCRIPTION
**Pull request by aikaniel**
This PR adds a method to the sprint model that returns true if the current day is within the sprint start and end period, and also returns false if either start or end date on the sprint are set to null.

Refactors the sprint chart service to use this method.

### CODE REVIEW

File: app/Models/Sprint.php [hunk 1]
```diff
@@ -81,4 +82,16 @@ public function getLastUpdatedDate(): Carbon
     {
         return $this->lastUpdated;
     }
+
+    public function containsToday(): Bool
+    {
+        if (empty($this->startDate)  or empty($this->endDate)) {
```
Comment (jtowe1) [inline app/Models/Sprint.php:88]: I'd use the getters here getStartDate() and getEndDate().  Also, does `or` actually work?  I'd replace it with `||` even if it does.

File: app/Models/Sprint.php [hunk 2]
```diff
@@ -81,4 +82,16 @@ public function getLastUpdatedDate(): Carbon
     {
         return $this->lastUpdated;
     }
+
+    public function containsToday(): Bool
```
Comment (jtowe1) [inline app/Models/Sprint.php:86]: I think it needs to be lowercase `bool` vs `Bool`
```suggestion
    public function containsToday(): bool
```

File: app/Models/Sprint.php (hunk 1 above)
Comment (aikaniel) [reply app/Models/Sprint.php:88]: Good catch! I'll update those to the getters...

For the or logic, yes it does seem to work. I wasn't sure at the time which was valid php at the time, and a quick google search I found told me this:
> The || and or operators return the same result. The only difference between the || and or operators are their precedences. The or operator has higher precedence than the || operator.

Considering that I just picked one semi-randomly! If you prefer `||` I don't have a problem switching to that!

File: app/Models/Sprint.php (hunk 2 above)
Comment (aikaniel) [reply app/Models/Sprint.php:86]: Ugh, you're right but nothing seemed to complain like I would have expected! I'll change that.

File: app/Models/Sprint.php [hunk 3]
```diff
@@ -81,4 +82,16 @@ public function getLastUpdatedDate(): Carbon
     {
         return $this->lastUpdated;
     }
+
+    public function containsToday(): bool
+    {
+        if (empty($this->getStartDate()) || $this->getEndDate()) {
+            return false;
+        }
```
Comment (jtowe1) [inline app/Models/Sprint.php:90]: I think the '||' check has to be explicit on both side.
```suggestion php
        if (empty($this->getStartDate()) || empty($this->getEndDate())) {
            return false;
        }
```

Comment (jtowe1) [reply app/Models/Sprint.php:90]: Also, this is super nitpicky of me, but can we do the else and both brackets on the same line like
```php
        } else {
````

File: app/Services/SprintChartService.php [hunk 4]
```diff
@@ -50,19 +49,16 @@ private function getSprintById(int $sprintId): Sprint
...
 
-            if ($period->contains(Carbon::now())) {
+        if (Carbon::now()->isWeekday()) {
+            if ($sprint->containsToday()) {
                 $issuesResponse = $this->getIssuesFromJiraBySprintId($sprintId);
                 $day = $this->buildNewDayFromJiraResponse($issuesResponse, $sprintId);
                 $this->dayRepository->save($day);
+                $sprint = $this->sprintRepository->loadById($sprintId);
             }
         }
```
Comment (jtowe1) [inline app/Services/SprintChartService.php:60]: You could combine these and decrease the indent level
```suggestion
        if (Carbon::now()->isWeekday() && $sprint->containsToday()) {
            $issuesResponse = $this->getIssuesFromJiraBySprintId($sprintId);
            $day = $this->buildNewDayFromJiraResponse($issuesResponse, $sprintId);
            $this->dayRepository->save($day);
            $sprint = $this->sprintRepository->loadById($sprintId);
        }
```
OR if you want to get real fancy you could make a function on `Sprint` called `shouldUpdateToday` like
```php
public function shouldUpdateToday(): bool
{
    return Carbon::now()->isWeekday() && $this->containsToday();
}
```
and then change the if in the SprintChartService to be
```php
        if ($sprint->shouldUpdateToday()) {
            $issuesResponse = $this->getIssuesFromJiraBySprintId($sprintId);
            $day = $this->buildNewDayFromJiraResponse($issuesResponse, $sprintId);
            $this->dayRepository->save($day);
            $sprint = $this->sprintRepository->loadById($sprintId);
        }
```

File: app/Models/Sprint.php (hunk 3 above)
Comment (aikaniel) [reply app/Models/Sprint.php:90]: Updating this to:

```php
    public function containsToday(): bool
    {
        $period = $this->sprintPeriod();
        return !empty($period) ? $period->contains(Carbon::now()) : false;
    }

    public function sprintPeriod(): ?CarbonPeriod
    {
        if (empty($this->getStartDate()) || empty($this->getEndDate())) {
            return null;
        }
        return CarbonPeriod::create($this->getStartDate(), $this->getEndDate());
    }
```

This should resolve both issues, but please feel free to review again!

File: app/Services/SprintChartService.php (hunk 4 above)
Comment (aikaniel) [reply app/Services/SprintChartService.php]: As we discussed, the weekend isn't part of the actual sprint so I've followed the second suggestion but reconciled the changes to just the `containsToday()` method. I did so by adding a guard:
```php
    public function containsToday(): bool
    {
        if (Carbon::now()->isWeekend()) {
            return false;
        }
        $period = $this->sprintPeriod();
        return !empty($period) ? $period->contains(Carbon::now()) : false;
    }
```
Then removed the weekday check in the SprintChartService.

**Review by jtowe1 — APPROVED** [review]
Nice work!

The migration correctly sets the column types

```mysql> describe sprint;
+-------------------+--------------+------+-----+---------+-------+
| Field             | Type         | Null | Key | Default | Extra |
+-------------------+--------------+------+-----+---------+-------+
| id                | int          | NO   | PRI | NULL    |       |
| name              | text         | NO   |     | NULL    |       |
| board_id          | int          | NO   |     | NULL    |       |
| length            | int          | NO   |     | NULL    |       |
| start_date        | datetime     | YES  |     | NULL    |       |
| end_date          | datetime     | YES  |     | NULL    |       |
| total_points      | varchar(255) | NO   |     | NULL    |       |
| total_goal_points | varchar(255) | NO   |     | NULL    |       |
| created_at        | timestamp    | YES  |     | NULL    |       |
| updated_at        | timestamp    | YES  |     | NULL    |       |
+-------------------+--------------+------+-----+---------+-------+
10 rows in set (0.02 sec)
```

**Pull request merged**